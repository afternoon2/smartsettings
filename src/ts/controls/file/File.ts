import cuid from 'cuid';

import { Control } from '../control/Control';
import {
  FileControlProps, ControlListener, ControlListenerUpdate,
} from '../control/Control.types';
import { InternalState } from '../../root/RootNode';

import Styles from '../../../sass/file.sass';

export class File extends Control {
  public controlElement: HTMLInputElement;

  protected controlId: string = cuid();

  protected static template = (state: InternalState): string => `<label
      class="${Styles.file}"
      for="${state.controlId}"
    >
      Upload...
      <input
        id="${state.controlId}"
        type="file"
        title="${state.name}"
        ${state.disabled ? 'disabled' : ''}
        ${state.readonly ? 'readonly' : ''}
      />
    </label>
  `;

  constructor(props: FileControlProps) {
    super(props, File.template);
    this.controlElement = this.element.querySelector('[type="file"]') as HTMLInputElement;
    this.listeners.set('accept', this.onAcceptChange);
    this.bindActionListeners();
  }

  get accept(): string {
    return this.state.accept as string;
  }

  set accept(accept: string) {
    this.state.accept = accept;
  }

  get files(): FileList | null {
    return this.state.files as FileList;
  }

  private onAcceptChange: ControlListener = (update: ControlListenerUpdate) => {
    this.controlElement.setAttribute('accept', update.value as string);
  }

  private bindActionListeners() {
    this.controlElement.addEventListener('change', (event: Event) => {
      this.state.files = ((event as Event).target as HTMLInputElement).files as FileList;
    });
  }
}
