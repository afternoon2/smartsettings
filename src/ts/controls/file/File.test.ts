import { File } from './File';

let file: File;
let userListener: any;

describe('Name of the group', () => {
  beforeEach(() => {
    userListener = jest.fn();
    file = new File({
      id: 'id',
      name: 'name',
      options: {
        controlId: 'controlId',
      },
      parentElement: document.body,
      userListener,
    });
  });

  test('If it matches the snapshot', () => {
    expect(file.element).toMatchSnapshot();
  });

  test('Get and set accept attribute', () => {
    const accept: string = '.doc,.docx,application/msword,application/vnd.openxmlformats-officedocument.wordprocessingml.document';
    file.accept = accept;
    expect(file.accept).toBe(accept);
    expect(file.controlElement.getAttribute('accept')).toBe(accept);
  });
});