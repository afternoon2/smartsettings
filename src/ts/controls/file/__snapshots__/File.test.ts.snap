// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Name of the group If it matches the snapshot 1`] = `
<div
  class="control"
>
  <main
    class="control__body"
    id="id"
  >
    <p
      class="control__name"
      title="name"
    >
      name
    </p>
    <div
      class="control__content"
    >
      <label
        class="file"
        for="controlId"
      >
        
      Upload...
      
        <input
          id="controlId"
          title="name"
          type="file"
        />
        
    
      </label>
      
  
    </div>
  </main>
</div>
`;
